### ISSUE TITLE
New Source - NASA

### ISSUE THREAD

**Issue by Linuturk** [ifl0w/RandomWallpaperGnome3#20]
Would love to see the NASA Image of the Day as a source for this extension.

https://www.nasa.gov/multimedia/imagegallery/iotd.html

**Comment by ifl0w** [comment]
Hello, basically it would be possible but there is no API and no possibility to request a random image. Requesting the newest image of the day would obviously result in the same picture for the whole day regardless how often you change the wallpaper. That wallpaper would have the same problem as the Bing image of the day from #15.

**Comment by Linuturk** [comment]
I'm actually fine with it only changing daily, as their images are really great.

I did lookup the API docs for the NASA API: https://api.nasa.gov/index.html

Specifically this PDF which documents the image API: https://images.nasa.gov/docs/images.nasa.gov_api_docs.pdf

Note the `/search` endpoint that returns a number of images based on various search parameters. It wouldn't be too difficult to pick a random image out of that endpoint given it returns so many results.

Either way, I think they have a great library of images and I'd love to see them included.

**Comment by ifl0w** [comment]
Well you are right. I did not look that up quiet well.
It would not be difficult to use the API but since there are nearly endless sources for wallpaper I am thinking of a more generic way of using such sources. 

For example a JSON API source where you have to provide the API URL and the JSON member of the result that contains the picture url you want to use as wallpaper. This would support the Bing picture of the day, the NASA API and probably many more sources but would require more technical know how from the user.

**Comment by Linuturk** [comment]
I'd have to see a more fleshed out example. Not sure how you'd handle the wide variety of possible JSON responses. Or the ability to choose something randomly from a JSON array. Plus API keys and other authentication mechanisms. All in all, I'd love to see it. If you could point me in the right direction, I might be able to put together a PR for the NASA provider.

**Comment by ifl0w** [comment]
It is not that hard to add a additional source. You'd currently need to do the following steps:

* add the select option to the settings with Glade
* add a new value to the enum in the settings schema and recompile the schema
* optionally add additinal source settings with Glade and to the schema.
* connect the schame to the UI in the `prefs.js`
* add a subclass of the BaseAdapter in the `sourceAdapter.js`
* add the new subclass to the adapter switch statement in the `wallpaperController.js`

This process could be made simpler but I didn't find it that important to think about it and make it cleaner. 

Basically the NASA source would be copy and paste of the desktopper source but I'd like to discuss the Generic API in #22.

**Comment by Linuturk** [comment]
Closing in favor of #22
